### PR TITLE
Add `dappSuggestedGasFees` to transaction

### PR DESCRIPTION
**Pull request by OGPoyraz**
## Explanation

This PR aims to add `dappSuggestedGasFees` if transaction creation origin is not MetaMask.

### `@metamask/transaction-controller`

- **ADDED**: Add `dappSuggestedGasFees` to transaction meta

## Checklist

- [X] I've updated the test suite for new or updated code as appropriate
- [X] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [X] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/transaction-controller/src/TransactionController.ts [hunk 1]
```diff
@@ -1292,6 +1303,40 @@ export class TransactionController extends BaseController<
 
     return Common.custom(customChainParams);
   }
+
+  private generateDappSuggestedGasFee(
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:1307]:
```suggestion
  private generateDappSuggestedGasFees(
```

File: packages/transaction-controller/src/TransactionController.ts [hunk 2]
```diff
@@ -1292,6 +1303,40 @@ export class TransactionController extends BaseController<
...
+    transaction: Transaction,
+    origin?: string,
+  ) {
+    let dappSuggestedGasFees: DappSuggestedGasFees | undefined;
+    // If we are dealing with a transaction suggested by a dapp and not
+    // an internally created metamask transaction, we need to keep record of
+    // the originally submitted gasParams.
+    if (typeof origin === 'string' && origin !== ORIGIN_METAMASK) {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.ts:1315]: I appreciate you've copied this exactly from the extension, but is it worth simplifying the code whenever possible to make any future maintenance easier.

Could we add some early returns and remove the `typeof` checks here to end up with something like:

```
  if (!origin || origin == ORIGIN_METAMASK) {
    return undefined;
  }

  const { gasPrice, maxFeePerGas, maxPriorityFeePerGas, gas } = transaction;

  if (!gasPrice && !maxFeePerGas && !maxPriorityFeePerGas && !gas) {
    return undefined;
  }

  const dappSuggestedGasFees = {};

  if (gasPrice) {
    dappSuggestedGasFees.gasPrice = gasPrice;
  } else if (maxFeePerGas || maxPriorityFeePerGas) {
    dappSuggestedGasFees.maxFeePerGas = maxFeePerGas;
    dappSuggestedGasFees.maxPriorityFeePerGas = maxPriorityFeePerGas;
  }

  if (gas) {
    dappSuggestedGasFees.gas = gas;
  }

  return dappSuggestedGasFees;
```

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 3]
```diff
@@ -682,6 +683,54 @@ describe('TransactionController', () => {
       expect(controller.state.transactions[0].status).toBe(
         TransactionStatus.unapproved,
       );
+      expect(
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:686]: Should we avoid duplicating assertions so we can limit the failed tests if a "feature" breaks in future?

File: packages/transaction-controller/src/TransactionController.test.ts [hunk 4]
```diff
@@ -682,6 +683,54 @@ describe('TransactionController', () => {
...
+        controller.state.transactions[0]?.dappSuggestedGasFees?.gasPrice,
+      ).toBe('0x1');
+    });
+
+    it('adds dappSuggestedGasFees to transaction if origin is not MM', async () => {
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/TransactionController.test.ts:691]: Should we have similar tests for the below so we explicitly cover all the branches?

- No Origin
- Origin Is MetaMask
- Gas Price
- Max Fee Per Gas & Max Priority Fee Per Gas

We could create a new describe to group all these.

File: packages/transaction-controller/src/TransactionController.ts (hunk 2 above)
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.ts]: Thanks for the suggestion, done in https://github.com/MetaMask/core/pull/1617/commits/fc0409f783f018d390edf376a587f2006b62335b

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 3 above)
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.test.ts]: This should be addressed in latest commit

File: packages/transaction-controller/src/TransactionController.test.ts (hunk 4 above)
Comment (OGPoyraz) [reply packages/transaction-controller/src/TransactionController.test.ts]: Done

File: packages/transaction-controller/src/types.ts [hunk 5]
```diff
@@ -149,3 +150,18 @@ export interface RemoteTransactionSource {
     request: RemoteTransactionSourceRequest,
   ) => Promise<TransactionMeta[]>;
 }
+
+/**
+ * We attach an object to transactions proposed by dapps to show the values
```
[the remark this answers is not in the document]
Comment (OGPoyraz) [reply packages/transaction-controller/src/types.ts]: Done

File: packages/transaction-controller/src/types.ts [hunk 6]
```diff
@@ -1,50 +1,105 @@
...
+ * Representation of transaction metadata.
  */
 export type TransactionMeta =
   | ({
       status: Exclude<TransactionStatus, TransactionStatus.failed>;
     } & TransactionMetaBase)
   | ({ status: TransactionStatus.failed; error: Error } & TransactionMetaBase);
 
+/**
+ * Representation of transaction metadata.
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/types.ts:13]:
```suggestion
 * Information about a single transaction such as status and block number.
```

File: packages/transaction-controller/src/types.ts [hunk 7]
```diff
@@ -74,32 +129,72 @@ export enum WalletDevice {
...
+
+  /**
+   * Value associated with this transaction.
+   */
   value?: string;
+
+  /**
+   * Maximum fee per gas for this transaction.
+   */
   maxFeePerGas?: string;
```
Comment (matthewwalsh0) [inline packages/transaction-controller/src/types.ts:183]: We're alphabetical up to here, shall we take the opportunity to order the rest?

Comment (OGPoyraz) [reply packages/transaction-controller/src/types.ts:178]: Done

Comment (OGPoyraz) [reply packages/transaction-controller/src/types.ts:178]: Dunno how did I missed it heh

**Review by matthewwalsh0 — APPROVED** [review]

**Review by vinistevam — APPROVED** [review]

**Pull request merged**